src/ruby/puppetserver-lib/puppet/server/ast_compiler.rb

```diff
@@ -24,26 +24,79 @@ def self.compile(compile_options)
+            'transport' => 'redacted',
+            'transports' => {
+              'redacted' => 'redacted'
+            }
+          }
+          bolt_inv = Bolt::ApplyInventory.new(fake_config)
+          Puppet.override(bolt_inventory: bolt_inv) do
+            Puppet.lookup(:pal_current_node).trusted_data = compile_options['trusted_facts']['values']
+            #TODO: get rid of this debug logging...
+            Puppet.warning("Compile Options")
+            Puppet.warning("#{compile_options}")
+            # This compiler has been configured with a node containing
+            # the requested environment, facts, and variables, and is used
+            # to compile a catalog in that context from the supplied AST.
+            pal.with_catalog_compiler do |compiler|
+              #CODEREVIEW: Do we need to make these configurable in the request?
+              Puppet[:strict] = :warning
+              Puppet[:strict_variables] = false
+              vars = Puppet::Pops::Serialization::FromDataConverter.convert(compile_options['variables']['values'])
+
+              # If fact conflicts with plan var name, fact wins, so delete plan var
+              # CODEREVIEW: logging is strange for this. For vars that are overridden, that would
+              # need to happen in orch logs, for facts it would happen here. Is it ok to just
+              # silently overwrite?
+              compile_options['facts']['values'].keys.each {|fact_name| vars.delete(fact_name)}
+              Puppet.warning("FACTS KEYS #{compile_options['facts']['values'].keys}")
+              pal.send(:add_variables, compiler.send(:topscope), vars)
+
+              # We have to parse the AST inside the compiler block, because it
+              # initializes the necessary type loaders for us.
+              ast = Puppet::Pops::Serialization::FromDataConverter.convert(code)
+              unless ast.is_a?(Puppet::Pops::Model::Program)
+                Puppet.warning("AST IS A: Puppet::Pops::Model::Program")
+                # Node definitions must be at the top level of the apply block.
+                # That means the apply body either a) consists of just a
+                # NodeDefinition, b) consists of a BlockExpression which may
+                # contain NodeDefinitions, or c) doesn't contain NodeDefinitions.
+                definitions = if ast.is_a?(Puppet::Pops::Model::BlockExpression)
+                                ast.statements.select { |st| st.is_a?(Puppet::Pops::Model::NodeDefinition) }
+                              elsif ast.is_a?(Puppet::Pops::Model::NodeDefinition)
+                                [ast]
+                              else
+                                []
+                              end
+                ast = Puppet::Pops::Model::Factory.PROGRAM(ast, definitions, ast.locator).model
+              end
+              compiler.evaluate(ast)
+              compiler.instance_variable_get(:@internal_compiler).send(:evaluate_ast_node)
```

Comment:
Why are we doing this? afaict, PAL only exists to support Bolt. Seems weird to have an interface's only consumer have to meta-program around the API.

Reply:
I did not consider that. I just used what was already in bolt. https://github.com/puppetlabs/bolt/blob/bafdcb788c5227244991ac721917732a4008ed76/lib/bolt/catalog.rb#L130

Reply:
https://github.com/puppetlabs/bolt/commit/09aed28c46a0d03530a50a4fd5825d30b43a78be Looks like it supports node definitions in apply blocks.